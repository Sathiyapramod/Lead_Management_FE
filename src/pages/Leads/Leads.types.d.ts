export interface LeadList {
    id: number;
    lead_name: string;
    rest_name: string;
    rest_addr1: string;
    rest_addr2: string;
    phone: string;
    mgr_id: number;
    lead_status?: boolean;
    call_freq: "weekly" | "daily";
    last_call_date?: string;
    orders_placed?: number;
    orders_done?: number;
    created_at: string;
    updated_at: string;
}
