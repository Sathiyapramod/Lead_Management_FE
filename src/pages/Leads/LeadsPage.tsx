import React, { useEffect,useState } from 'react';
import { useNavigate } from 'react-router-dom';

import Button from '../../components/Button';
import Card from '../../components/Card';
import Table from '../../components/custom/Table/Table';
import Heading from '../../components/Heading';
import SearchBar from '../../components/SearchBar';
import { useAppDispatch, useTypedSelector } from '../../store';
import { fetchLeadLists } from '../../store/reducers/leads';
import { fetchLeads } from '../../store/reducers/performance';
import { LeadColumns } from '../../utils/constants';
import { title_headings } from '../../utils/headings';

function Leads(): React.ReactNode {
  const dispatch = useAppDispatch();
  const navigate = useNavigate();

  const [searchName, setName] = useState<string>('');
  const [offset, setOffset] = useState<number>(0);

  useEffect(() => {
    dispatch(
      fetchLeadLists({ offset: String(offset), limit: '10', searchName }),
    );
    dispatch(fetchLeads());
  }, [offset, searchName]);

  const { leadList } = useTypedSelector((state) => state.leads);
  const { analytics } = useTypedSelector((state) => state);

  const onClick = () => {
    if (window.localStorage.getItem('role') === 'KAM')
      navigate('/leads/create');
    else return;
  };

  return (
    <div className="p-[55px]">
      <div className="flex justify-between items-start">
        <Card
          title={title_headings.LEADS}
          count={analytics.leadList?.count ?? 0}
          active={analytics.leadList?.active ?? 0}
          pending={analytics.leadList?.pending ?? 0}
          activeTag="Active"
          pendingTag="Pending"
        />
        <div className="text-right">
          <SearchBar
            onChange={(e) => setName(e.target.value)}
            placeholder={'Search Leads'}
          />
          <Button
            content="+ Create Lead"
            theme="dark"
            classname="rounded-lg mt-[15px]"
            onClick={onClick}
          />
        </div>
      </div>
      <Heading content={title_headings.LEADS} classname="my-[40px]" />
      <Table
        columns={LeadColumns}
        data={leadList}
        setOffset={setOffset}
        name={title_headings.LEADS}
      />
    </div>
  );
}

export default Leads;
