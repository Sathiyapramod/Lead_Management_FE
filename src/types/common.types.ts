export interface LeadsQuery {
    call_freq: string;
    last_call_date: string;
    limit: string;
    offset: string;
    mgr_id: string;
    today: string;
}
