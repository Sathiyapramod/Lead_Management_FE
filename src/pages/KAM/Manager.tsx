import React, { useEffect, useState } from "react";
import { toast } from "sonner";

import Card from "../../components/Card";
import Table from "../../components/custom/Table/Table";
import { BaseTable } from "../../components/custom/Table/Tables.types";
import Heading from "../../components/Heading";
import SearchBar from "../../components/SearchBar";
import API from "../../services/api";
import { MgrColumns } from "../../config/constants";
import { title_headings } from "../../config/headings";
import { ManagersList } from "./Manager.types";

interface Managers extends BaseTable {
  managers: ManagersList[];
}

function Manager(): React.ReactNode {
  const [searchName, setName] = useState<string>("");
  const [mgrs, setMgrs] = useState<Managers>({
    count: 0,
    active: 0,
    pending: 0,
    completed: 0,
    managers: [],
  });
  const [offset, setOffset] = useState<number>(0);

  const getList = async () => {
    try {
      const { data, status } = await API.getManagers({
        limit: "10",
        offset: String(offset),
        searchName,
      });
      if (status !== 200) {
        toast.error("Error while fetching the Leads");
      } else {
        setMgrs(data);
      }
    } catch (err) {
      console.log(err);
      toast.error("Error");
    }
  };

  useEffect(() => {
    getList();
  }, [offset, setName]);

  return (
    <div className="p-[55px]">
      <div className="flex justify-between items-start">
        <Card
          title={title_headings.KAM}
          count={mgrs?.count ?? 0}
          active={mgrs?.active ?? 0}
          pending={mgrs?.pending ?? 0}
          activeTag="Active"
          pendingTag="Pending"
        />
        <div className="text-right">
          <SearchBar
            onChange={(e) => setName(e.target.value)}
            placeholder={"Search Managers"}
          />
        </div>
      </div>
      <Heading content={title_headings.KAM} classname="my-[40px]" />
      <Table
        columns={MgrColumns}
        data={mgrs}
        setOffset={setOffset}
        name={title_headings.KAM}
      />
    </div>
  );
}

export default Manager;
