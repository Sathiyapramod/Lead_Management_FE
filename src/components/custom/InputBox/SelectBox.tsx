import React from "react";

import { ManagersList } from "../../../pages/KAM/Manager.types";
import { LeadList } from "../../../pages/Leads/Leads.types";

interface AppSelectBox {
    label: string;
    list?: Partial<ManagersList>[];
    leads?: Partial<LeadList>[];
    value: string;
    defaultValue: string;
    listName: "managers" | "leads";
    onChange: (e: React.ChangeEvent<HTMLSelectElement>) => void;
}
function SelectBox({ label, value, list, onChange, defaultValue, listName, leads }: AppSelectBox) {
    return (
        <div className="mb-4 lg:basis-1/5">
            <label className="block text-left text-sm font-medium text-gray-70">{label}</label>
            <select
                value={value}
                onChange={onChange}
                className="mt-1 block p-2 border-b-[0.5px] border-light-black focus:border-blue-300 focus:outline-none max-lg:w-[325px] min-[320px]:w-full max-md:w-full"
                disabled={false}
                required
            >
                <option value="select">{defaultValue}</option>
                {listName === "managers" &&
                    list &&
                    list.map((eachItem) => (
                        <option value={eachItem.id} key={eachItem.id}>
                            {eachItem.mgr_name}
                        </option>
                    ))}
                {listName === "leads" &&
                    leads &&
                    leads.map((eachItem) => (
                        <option value={eachItem.id} key={eachItem.id}>
                            {eachItem.lead_name}
                        </option>
                    ))}
            </select>
        </div>
    );
}

export default SelectBox;
