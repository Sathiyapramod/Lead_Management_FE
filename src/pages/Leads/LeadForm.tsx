import React, { useEffect, useState } from "react";
import { useParams, useNavigate } from "react-router-dom";
import Button from "../../components/Button";
import Heading from "../../components/Heading";
import API from "../../services/api";
import { toast } from "sonner";
import { useAppDispatch } from "../../store";
import { resetState } from "../../store/reducers/leads";
import { LeadList } from "./LeadsPage";
import TextBox from "../../components/custom/InputBox/TextBox";
import SelectBox from "../../components/custom/InputBox/SelectBox";
import { ManagersList } from "../KAM/Manager";
import Stepper from "../../components/custom/Stepper/Stepper";
import ToggleButton from "../../components/custom/ToggleButton";

export interface AppLeadForm {
    sub: "create" | "edit";
}

function LeadForm({ sub }: AppLeadForm) {
    const dispatch = useAppDispatch();
    const { id } = useParams();
    const navigate = useNavigate();
    const [mgrs, setMgrs] = useState<ManagersList[]>([]);
    const [formData, setFormData] = useState<LeadList>({
        id: 0,
        lead_name: "",
        rest_name: "",
        rest_addr1: "",
        rest_addr2: "",
        phone: "",
        lead_status: false,
        mgr_id: "",
        call_freq: "weekly",
        created_at: "",
        updated_at: "",
    });

    const getManagers = async () => {
        try {
            const { data, status } = await API.getManagers({});
            if (status !== 200) {
                toast.error("Error while fetching the Leads");
            } else {
                setMgrs(data.managers);
            }
        } catch (err) {
            toast.error("Error");
        }
    };

    const getLeadById = async (id: number) => {
        try {
            const { data, status } = await API.getLeadById(id);
            if (status !== 200) {
                toast.error("Error while fetching the Leads");
            } else {
                setFormData(data);
            }
        } catch (err) {
            toast.error("Error");
        }
    };

    useEffect(() => {
        if (id) getLeadById(Number(id));
    }, [id]);

    useEffect(() => {
        getManagers();
    }, []);

    const updateField = (field: string) => (value: string) => {
        setFormData((prevData) => ({
            ...prevData,
            [field]: value,
        }));
    };

    const createLead = async () => {
        try {
            const { status } = await API.createLead({
                ...formData,
                mgr_id: Number(formData.mgr_id),
                lead_status: true,
            });
            if (status !== 201) {
                toast.error("Error while creating Leads");
            } else {
                dispatch(resetState());
                navigate("/leads");
                toast.success("Lead Created successfully");
            }
        } catch (err) {
            toast.error("Error");
        }
    };

    const updateLead = async () => {
        try {
            const { status } = await API.createLead(formData);
            if (status !== 200) {
                toast.error("Error while creating Leads");
            } else {
                navigate(-1);
                dispatch(resetState());
                toast.success("Lead Created successfully");
            }
        } catch (err) {
            toast.error("Error");
        }
    };

    const handleSubmit = (e: React.FormEvent) => {
        e.preventDefault();
        if (sub === "create") createLead();
        else updateLead();
    };

    return (
        <div className="lg:mx-[55px] bg-white">
            <Heading
                content={`${sub === "create" ? "Create" : "Edit"} Lead`}
                classname="px-8 pt-8 my-[40px] max-md:px-5"
            />
            {sub === "edit" && <Stepper status={formData.lead_status} />}
            <form
                onSubmit={handleSubmit}
                className="mx-auto p-4 max-md:mx-5 md:w-full lg:w-full lg:p-[30px] md:block lg:flex lg:flex-row lg:justify-start lg:flex-wrap gap-8 relative"
            >
                <TextBox
                    label={"Lead Name"}
                    value={formData.lead_name}
                    onChange={(e) => updateField("lead_name")(e.target.value)}
                />
                <TextBox
                    label={"Restaurant Name"}
                    value={formData.rest_name}
                    onChange={(e) => updateField("rest_name")(e.target.value)}
                />
                <TextBox
                    label={"Restaurant Address 1"}
                    value={formData.rest_addr1}
                    onChange={(e) => updateField("rest_addr1")(e.target.value)}
                />
                <TextBox
                    label={"Restaurant Address 2"}
                    value={formData.rest_addr2}
                    onChange={(e) => updateField("rest_addr2")(e.target.value)}
                />
                <TextBox
                    label={"Phone"}
                    value={formData.phone}
                    onChange={(e) => updateField("phone")(e.target.value)}
                />
                <SelectBox
                    value={formData.mgr_id.toString()}
                    defaultValue="Select Manager"
                    label="Manager"
                    list={mgrs}
                    onChange={(e) => updateField("mgr_id")(e.target.value)}
                />
                <div className="mb-4 lg:basis-1/5">
                    <label className="block text-left text-sm font-medium text-gray-70">
                        Call Frequency
                    </label>
                    <select
                        value={formData.call_freq}
                        onChange={(e) => updateField("call_freq")(e.target.value)}
                        className="mt-1 block p-2 border-b-[0.5px] border-light-black focus:border-blue-300 focus:outline-none max-lg:w-[325px] min-[320px]:w-full max-md:w-full"
                        required
                    >
                        <option value="daily">Daily</option>
                        <option value="weekly">Weekly</option>
                    </select>
                </div>
                {sub === "edit" && (
                    <ToggleButton
                        value={formData.lead_status}
                        onChange={() =>
                            setFormData((pv) => {
                                const { lead_status } = pv;
                                return { ...pv, lead_status: !lead_status };
                            })
                        }
                        label="Lead Status"
                    />
                )}
                <Button
                    type="submit"
                    content="Submit"
                    theme="dark"
                    classname="absolute inset bottom-[35px] right-[25px] max-md:relative max-md:mt-[50px] max-md:mx-3"
                />
            </form>
        </div>
    );
}

export default LeadForm;
