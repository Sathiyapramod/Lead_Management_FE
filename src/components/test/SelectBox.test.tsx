import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import SelectBox from "../custom/InputBox/SelectBox";
import { ManagersList } from "../../pages/KAM/Manager.types";
import { LeadList } from "../../pages/Leads/Leads.types";

describe("TextBox Component", () => {
    const mockManagersList: Partial<ManagersList>[] = [
        { id: 1, mgr_name: "Manager One" },
        { id: 2, mgr_name: "Manager Two" },
    ];

    const mockLeadsList: Partial<LeadList>[] = [
        { id: 1, lead_name: "Lead One" },
        { id: 2, lead_name: "Lead Two" },
    ];
    test("renders the text box ", () => {
        render(
            <SelectBox
                label="Select Manager"
                value="select"
                defaultValue="Select a manager"
                listName="managers"
                list={mockManagersList}
                onChange={() => {}}
            />
        );

        const labelElement = screen.getByLabelText(/select manager/i);
        expect(labelElement).toBeInTheDocument();
    });

    test("displays the default value correctly", () => {
        render(
            <SelectBox
                label="Select Manager"
                value="select"
                defaultValue="Select a manager"
                listName="managers"
                list={mockManagersList}
                onChange={() => {}}
            />
        );

        const selectElement = screen.getByRole("combobox");
        expect(selectElement).toHaveValue("select");
    });

    test("renders options based on the managers list", () => {
        render(
            <SelectBox
                label="Select Manager"
                value="select"
                defaultValue="Select a manager"
                listName="managers"
                list={mockManagersList}
                onChange={() => {}}
            />
        );

        const optionElements = screen.getAllByRole("option");
        expect(optionElements).toHaveLength(mockManagersList.length + 1); // +1 for the default option
        expect(screen.getByText("Manager One")).toBeInTheDocument();
        expect(screen.getByText("Manager Two")).toBeInTheDocument();
    });

    test("renders options based on the leads list", () => {
        render(
            <SelectBox
                label="Select Lead"
                value="select"
                defaultValue="Select a lead"
                listName="leads"
                leads={mockLeadsList}
                onChange={() => {}}
            />
        );

        const optionElements = screen.getAllByRole("option");
        expect(optionElements).toHaveLength(mockLeadsList.length + 1); // +1 for the default option
        expect(screen.getByText("Lead One")).toBeInTheDocument();
        expect(screen.getByText("Lead Two")).toBeInTheDocument();
    });

    test("calls onChange handler when selecting an option", () => {
        const handleChange = jest.fn();
        render(
            <SelectBox
                label="Select Manager"
                value="select"
                defaultValue="Select a manager"
                listName="managers"
                list={mockManagersList}
                onChange={handleChange}
            />
        );

        const selectElement = screen.getByRole("combobox");
        fireEvent.change(selectElement, { target: { value: "1" } });

        expect(handleChange).toHaveBeenCalledTimes(1);
    });

    test("renders with the correct value", () => {
        render(
            <SelectBox
                label="Select Manager"
                value="1"
                defaultValue="Select a manager"
                listName="managers"
                list={mockManagersList}
                onChange={() => {}}
            />
        );
    });

    test("snapshot select box", () => {
        const { asFragment } = render(
            <SelectBox
                label="Select Manager"
                value="1"
                defaultValue="Select a manager"
                listName="managers"
                list={mockManagersList}
                onChange={() => {}}
            />
        );

        expect(asFragment).toMatchSnapshot();
    });
});
