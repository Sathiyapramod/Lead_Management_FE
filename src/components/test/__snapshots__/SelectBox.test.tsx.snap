// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextBox Component snapshot select box 1`] = `[Function]`;
